server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -352,19 +352,18 @@ private synchronized void onMergeFailure(Exception exc) {
                 circuitBreakerBytes = 0;
             }
             failure.compareAndSet(null, exc);
-            MergeTask task = runningTask.get();
-            runningTask.compareAndSet(task, null);
-            onPartialMergeFailure.accept(exc);
-            List<MergeTask> toCancels = new ArrayList<>();
+            final MergeTask task = runningTask.getAndSet(null);
```

Comment:
This one seems to me to be the most likely place where something could go wrong and we should really guard against a `cancel` throwing in the loop.